pkg/controller/meta_pod_controller.go

```diff
@@ -0,0 +1,231 @@
+			memberAddress = arr[0]
+		} else {
+			log.Error(fmt.Errorf("unexpected member format %s", member.Address.Host), "")
+			return consts.MetaRoleUnknown, false
+		}
+		if member.IsLeader && podName == memberAddress && port == uint(member.Address.Port) {
+			return consts.MetaRoleLeader, true
+		}
+	}
+	return consts.MetaRoleFollower, true
+}
+
+// getMetaPort returns the service port of the pod. Returns true if operation was successful.
+func getMetaPort(pod *corev1.Pod) (uint, bool) {
+	for _, container := range pod.Spec.Containers {
+		if container.Name == "meta" {
+			for _, containerPort := range container.Ports {
+				if containerPort.Name == consts.PortService {
+					return uint(containerPort.ContainerPort), true
+				}
+			}
+		}
+	}
+	return 0, false
+}
+
+// getPodIP returns the pod IP of the pod. Returns true if operation was successful.
+func getPodIP(pod *corev1.Pod) (string, bool) {
+	ip := pod.Status.PodIP
+	return ip, ip != ""
+}
+
+// getOtherMetPods returns all other meta pods of the same RW instance.
+func (mpc *MetaPodController) getOtherMetPods(metaPod *corev1.Pod, ctx context.Context) ([]corev1.Pod, error) {
+	log := log.FromContext(ctx)
+
+	// get other meta pods of this RW instance
+	otherMetaPods := &corev1.PodList{}
+	labelSet := map[string]string{consts.LabelRisingWaveComponent: consts.ComponentMeta}
+	if rwInstance, ok := metaPod.ObjectMeta.Labels[consts.LabelRisingWaveName]; ok {
+		labelSet[consts.LabelRisingWaveName] = rwInstance
+	} else {
+		log.Error(fmt.Errorf("unable to retrieve risingwave name from pod"), "")
+	}
+	listOptions := client.ListOptions{LabelSelector: labels.SelectorFromSet(labelSet)}
+	err := mpc.Client.List(context.Background(), otherMetaPods, &listOptions)
+	return otherMetaPods.Items, err
+}
+
+// Reconcile handles the pods of the meta service. Will add the metaLeaderLabel to the pods.
+func (mpc *MetaPodController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	defaultRequeue2sResult := ctrl.Result{RequeueAfter: time.Second * 2}
+	pendingRequeue10sResult := ctrl.Result{RequeueAfter: time.Second * 10}
+
+	// only reconcile when this is related to a meta pod
+	reqPod := &corev1.Pod{}
+	mpc.Get(ctx, req.NamespacedName, reqPod)
+	rwComponent, ok := reqPod.Labels[consts.LabelRisingWaveComponent]
+	if !ok || rwComponent != consts.ComponentMeta {
+		return defaultRequeue2sResult, nil
+	}
+
+	log := log.FromContext(ctx)
+
+	originalReqPod := reqPod.DeepCopy()
+	oldRole := reqPod.Labels[consts.LabelRisingWaveMetaRole]
+	timeoutCtx, cancel := context.WithTimeout(ctx, 3*time.Second)
+	defer cancel()
+
+	podIP, ok := getPodIP(reqPod)
+	if !ok {
+		log.Info("error when reconciling this pod. Assuming that pod is pending")
+		return pendingRequeue10sResult, nil
+	}
+	podPort, ok := getMetaPort(reqPod)
+	if !ok {
+		log.Error(fmt.Errorf("error when reconciling this pod"), "")
+		return defaultRequeue2sResult, nil
+	}
```

Comment:
They can be ignored and triggered when the Pod's status is changed. And for those Pods we can not extract a port from it, we just ignore them.